In [165]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas

# Import scikit-learn functions
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

# Ignore warnings about future changes in library and deprecated functions
import warnings
warnings.filterwarnings('ignore')

In [166]:
# Load the .csv data into a Pandas DataFrame with the index column set to the column with the coin_id
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display first ten rows of sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [167]:
# Generate summary statistics for the DataFrame
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [168]:
# Plot your data to see what's in your DataFrame before operations are performed
df_market_data.hvplot.line(
    width = 800,
    height = 400,
    rot = 90 # Rotates the x-axis labels to be vertical
) 

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [169]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data)
print(crypto_scaled[0:5]) 

[[ 0.50852937  0.49319307  0.77220043  0.23545963 -0.0674951  -0.35595348
  -0.25163688]
 [ 0.18544589  0.93444504  0.55869212 -0.05434093 -0.27348273 -0.11575947
  -0.19935211]
 [ 0.02177396 -0.70633685 -0.02168042 -0.06103015  0.00800452 -0.55024692
  -0.28206051]
 [-0.04076438 -0.81092807  0.24945797 -0.05038797 -0.37316402 -0.45825882
  -0.29554614]
 [ 1.19303608  2.00095907  1.76061001  0.54584206 -0.29120287 -0.49984776
  -0.27031695]]


In [170]:
# Create a DataFrame with the scaled data and the corresponding columns
crypto_transformed = pd.DataFrame(crypto_scaled, columns=["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d", 'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d',	'price_change_percentage_1y'])

# Copy the crypto names from the original data 
coins = list(df_market_data.index) # this returns a list of all the coin names from the original data

# Set the coin_id column as index
crypto_transformed["coin_id"] = coins  # create a new column in the transformed df containing the list of coins
crypto_transformed = crypto_transformed.set_index("coin_id") # set the df index to the new coin_id column

# Display first ten rows of transformed data
crypto_transformed.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Data.

In [171]:
# Create a list with the number of k-values from 1 to 11
k1 = list(range(1,12))

In [172]:
# Create an empty list to store the inertia values
inertia_list1 = []


# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i1 in k1:
    k_model = KMeans(n_clusters = i1, random_state = 1) # random_state is the seed for the random number generator
    k_model.fit(crypto_transformed) # with reference to the instructions, `crypto_transformed` is essentially: `df_market_data_scaled`
    inertia_list1.append(k_model.inertia_) # inertia is the sum of squared distances of samples to their closest cluster center

In [173]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k1, "inertia": inertia_list1}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df1 = pd.DataFrame(elbow_data)

In [174]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df1.hvplot.line(
    x = "k", 
    y = "inertia", 
    title = "Elbow Curve",
    xticks = k1
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for `k` = **4**. _After that point, the line becomes mostly horizontal, because more clusters only reduce the within-cluster sum of squares (WCSS) by a small amount._

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [176]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters = 4, random_state = 1) # random_state is the seed value

In [177]:
# Fit the K-Means model using the scaled data
model.fit(crypto_transformed)

KMeans(n_clusters=4, random_state=1)

In [178]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_4_clusters1 = model.predict(crypto_transformed) # .predict is the testing step

# Print the resulting array of cluster values.
k_4_clusters1

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1])

In [135]:
# Create a copy of the DataFrame
clusters_predicted = crypto_transformed.copy()

In [179]:
# Add a new column to the DataFrame with the predicted clusters
clusters_predicted["predicted_cluster"] = k_4_clusters1 

# Display first ten rows of predicted clusters
clusters_predicted.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  predicted_cluster  
coin_id                                                         
bitcoin                           -0.251637                  3  
ethereum                          -0.199352                  3  
tether                            -0.282061                  1  
ripple                            -0.295546                  1  
bitcoin-cash                      -0.270317                  3  
binancecoin                       -0.225533                  3  
chainlink                         -0.018284                  3  
cardano                           -0.155428                  3  
litecoin                          -0.292351                  3  
bitcoin-cash-sv                   -0.206029                  1

In [180]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_predicted.hvplot.scatter(
    x = "price_change_percentage_24h", # following naming instructions in the notebook, above, rather than the instructions in Bootcamp Spot
    y = "price_change_percentage_7d", # following naming instructions in the notebook, above, rather than the instructions in Bootcamp Spot
    by = "predicted_cluster",
    hover_cols = "coin_id" # this is the column that will be displayed when you hover over the data points
)

:NdOverlay   [predicted_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [181]:
# Create a PCA model instance and set `n_components=3`.
pca_instance = PCA(n_components = 3)

In [182]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
clusters_pca = pca_instance.fit_transform(crypto_transformed)

# View the first five rows of the DataFrame. 
clusters_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [184]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca_instance.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the dataset when all dimensional variance is reduced to three principal components is _0.3719856 + 0.34700813 + 0.17603793_ = **0.89503166**, or almost 90%. In other words, _even after PCA reduces dimensionality, it still captures a large amount of the original dataset's variance._

In [185]:
# Create a new DataFrame with the PCA data.
cluster_pca_df = pd.DataFrame(
    clusters_pca,
    columns = ["PC1", "PC2", "PC3"]
)

# Copy the crypto names from the original data
cluster_pca_df["coin_id"] = list(crypto_transformed.index)

# Set the coin_id column as index
cluster_pca_df = cluster_pca_df.set_index("coin_id")

# Display first ten rows of PCA data
cluster_pca_df.head(10)

PC1       PC2       PC3
coin_id                                      
bitcoin         -0.600667  0.842760  0.461595
ethereum        -0.458261  0.458466  0.952877
tether          -0.433070 -0.168126 -0.641752
ripple          -0.471835 -0.222660 -0.479053
bitcoin-cash    -1.157800  2.041209  1.859715
binancecoin     -0.516534  1.388377  0.804071
chainlink       -0.450711  0.517699  2.846143
cardano         -0.345600  0.729439  1.478013
litecoin        -0.649468  0.432165  0.600303
bitcoin-cash-sv -0.759014 -0.201200 -0.217653

---

### Find the Best Value for k Using the PCA Data

In [186]:
# Create a list with the number of k-values from 1 to 11
k2 = list(range(1,12))

In [187]:
# Create an empty list to store the inertia values
inertia_list2 = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i2 in k2:
    k_model = KMeans(n_clusters = i2, random_state = 1)
    k_model.fit(cluster_pca_df) # with reference to the instructions, `cluster_pca_df` is essentially: `df_market_data_pca`
    inertia_list2.append(k_model.inertia_)

In [188]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data2 = {"k": k2, "inertia": inertia_list2}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df2 = pd.DataFrame(elbow_data2)

In [189]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df2.hvplot.line(
    x = "k", 
    y = "inertia", 
    title = "Elbow Curve", 
    xticks = k2
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for `k` = **4**. _After that point, the line becomes mostly horizontal, because more clusters only reduce the within-cluster sum of squares (WCSS) by a small amount._


* **Question:** Does it differ from the best `k` value found using the original data?

  * **Answer:** No, the optimal value for `k` in both cases = **4**. This consistency in the optimal `k` value between the original scaled data and the PCA-reduced data provides additional confidence in the clustering. It suggests that the underlying structure of the data – in terms of how the data points group together – is maintained _even after the dimensionality reduction of PCA_ (i.e., "throwing away" some data).


### Cluster Cryptocurrencies with K-means Using the PCA Data

In [190]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters = 4, random_state = 1) # random_state is the seed value

In [191]:
# Fit the K-Means model using the PCA data
model.fit(cluster_pca_df) # .fit is the training step

KMeans(n_clusters=4, random_state=1)

In [192]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4_clusters2 = model.predict(cluster_pca_df)

# Print the resulting array of cluster values.
k_4_clusters2

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [193]:
# Create a copy of the DataFrame with the PCA data
copy_cluster_pca_df = cluster_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
copy_cluster_pca_df["predicted_cluster"] = k_4_clusters2

# Display first ten rows of predicted clusters
copy_cluster_pca_df.head(10)

PC1       PC2       PC3  predicted_cluster
coin_id                                                         
bitcoin         -0.600667  0.842760  0.461595                  1
ethereum        -0.458261  0.458466  0.952877                  1
tether          -0.433070 -0.168126 -0.641752                  0
ripple          -0.471835 -0.222660 -0.479053                  0
bitcoin-cash    -1.157800  2.041209  1.859715                  1
binancecoin     -0.516534  1.388377  0.804071                  1
chainlink       -0.450711  0.517699  2.846143                  1
cardano         -0.345600  0.729439  1.478013                  1
litecoin        -0.649468  0.432165  0.600303                  1
bitcoin-cash-sv -0.759014 -0.201200 -0.217653                  0

In [195]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
copy_cluster_pca_df.hvplot.scatter(
    x="PC1", # following naming instructions in the notebook, above, rather than the instructions in Bootcamp Spot
    y="PC2", # following naming instructions in the notebook, above, rather than the instructions in Bootcamp Spot
    by = 'predicted_cluster',
    hover_cols='coin_id' # this is the column that will be displayed when you hover over the data points
)

:NdOverlay   [predicted_cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [196]:
# Composite plot to contrast the Elbow curves
elbow_df1.hvplot.line(x = "k", y = "inertia", title = "Elbow Curve 1", xticks = k1) + elbow_df2.hvplot.line(x = "k", y = "inertia", title= "Elbow Curve 2", xticks = k2)

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [197]:
# Composite plot to contrast the clusters
clusters_predicted.hvplot.scatter( x = "price_change_percentage_24h", y = "price_change_percentage_7d", by = "predicted_cluster", hover_cols = "coin_id") + copy_cluster_pca_df.hvplot.scatter(x = "PC1", y = "PC2", by = 'predicted_cluster', hover_cols = 'coin_id')

:Layout
   .NdOverlay.I  :NdOverlay   [predicted_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predicted_cluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The impact of using PCA data results in tighter, more compact, less dispersed clusters. This is because the goal of PCA is to _capture as much of the variance in the data with fewer features (i.e., principal components)_. This can help clarify patterns in the data, especially when visualizing it. With less "noise", clusters are more condensed.